dtfs-central-api/src/services/state-machines/fee-record/fee-record.state-machine.ts

```diff
@@ -0,0 +1,67 @@
+import { FeeRecordEntity } from '@ukef/dtfs2-common';
+import { InvalidStateMachineTransitionError } from '../../../errors';
+import { FeeRecordRepo } from '../../../repositories/fee-record-repo';
+import { FeeRecordEvent } from './event/fee-record.event';
+import { handleFeeRecordAddAPaymentEvent } from './event-handlers';
+
+export class FeeRecordStateMachine {
+  private readonly feeRecord: FeeRecordEntity;
+
+  private constructor(feeRecord: FeeRecordEntity) {
+    this.feeRecord = feeRecord;
+  }
+
+  public static forFeeRecord(feeRecord: FeeRecordEntity): FeeRecordStateMachine {
+    return new FeeRecordStateMachine(feeRecord);
+  }
+
+  public static async forFeeRecordId(id: number): Promise<FeeRecordStateMachine> {
+    const feeRecord = await FeeRecordRepo.findOneByOrFail({ id });
+    return new FeeRecordStateMachine(feeRecord);
+  }
+
+  private handleInvalidTransition = ({ type: eventType }: FeeRecordEvent): never => {
+    const entityName = FeeRecordEntity.name;
+
+    throw InvalidStateMachineTransitionError.forEntity({
+      eventType,
+      entityName,
+      state: this.feeRecord.status,
+      entityId: this.feeRecord.id,
+    });
+  };
+
+  public async handleEvent(event: FeeRecordEvent): Promise<FeeRecordEntity> {
+    switch (this.feeRecord.status) {
+      case 'TO_DO':
+        switch (event.type) {
+          case 'ADD_A_PAYMENT':
+            return handleFeeRecordAddAPaymentEvent(this.feeRecord, event.payload);
```

Comment:
I think this should definitely be the job of another ticket, but there is some complexity around payments being added against multiple fee records... I wonder how that will work given the state transition will I assume happen as a group rather than individually...